pkg/ring/batch.go

```diff
@@ -30,19 +30,29 @@ type itemTracker struct {
 	failed4xx   atomic.Int32
 	failed5xx   atomic.Int32
 	remaining   atomic.Int32
-	err         atomic.Error
+	err4xx      atomic.Error
+	err5xx      atomic.Error
 }
 
 func (i *itemTracker) recordError(err error) int32 {
-	i.err.Store(err)
 
 	if status, ok := status.FromError(err); ok && status.Code()/100 == 4 {
+		i.err4xx.Store(err)
 		return i.failed4xx.Inc()
 	}
 
+	i.err5xx.Store(err)
 	return i.failed5xx.Inc()
 }
 
+func (i *itemTracker) getError() error {
+	if i.failed5xx.Load() > i.failed4xx.Load() {
+		return i.err5xx.Load()
+	}
+
+	return i.err4xx.Load()
```

Comment:
How about fetching the two values into local variables, so you don't make 4x atomic calls?

Reply:
Bryan, i am not sure i get your idea. They are 4 different values. 2 for amount of erros and 2 for the actual error. Current we can only return the last known error, the change is basically to hold one error for each family type.